_source/_posts/2022-04-26-webauthn-java.md

```diff
@@ -0,0 +1,713 @@
+                RegisteredCredential.builder()
+                    .credentialId(credential.getCredentialId())
+                    .userHandle(credential.getUser().getHandle())
+                    .publicKeyCose(credential.getPublicKey())
+                    .signatureCount(credential.getCount())
+                    .build())
+            .collect(Collectors.toSet());
+}
+```
+
+Yubico provides a `RelyingParty` object that's responsible for handling all authentication requests on the server, so the application constructs and provides that object to all controllers. Because WebAuthn credentials are scoped to only one website, the hostname, origin, and a display name for the website are all provided by the relying party.
+
+In the `AppApplication` class define the `RelyingParty` bean:
+
+```java
+@Bean
+public RelyingParty relyingParty(RegistrationRepository registrationRepository, 
+                                 WebAuthProperties properties) {
+    RelyingPartyIdentity rpIdentity = RelyingPartyIdentity.builder()
+        .id(properties.getHostName())
+        .name(properties.getDisplay())
+        .build();
+
+    return RelyingParty.builder()
+        .identity(rpIdentity)
+        .credentialRepository(registrationRepository)
+        .origins(properties.getOrigin())
+        .build();
+}
+```
+### Authentication controllers
+
+The web controller for the application creates routes and constructs the data needed for the client to make WebAuthn API requests by:
+
+- Constructing JSON strings used in the registration and authentication ceremonies.
+- Caching intermediary information necessary for the secure execution of a ceremony (the challenge nonce).
+- Passing information from the browser's credential creation ceremony to the data storage classes.
+
+Create an `AuthController` class:
+
+```java
+import com.yubico.webauthn.RelyingParty;
+import org.springframework.stereotype.Controller;
+
+@Controller
+public class AuthController {
+
+    private RelyingParty relyingParty;
+    private RegistrationService service;
+
+    AuthController(RegistrationService service, RelyingParty relyingPary) {
+        this.relyingParty = relyingPary;
+        this.service = service;
+    }
+    
+    @GetMapping("/")
+    public String welcome() {
+        return "index";
+    }
+    
+    @GetMapping("/login")
+    public String loginPage() {
+        return "login";
+    }
+
+    @GetMapping("/register")
+    public String registerUser(Model model) {
+        return "register";
+    }
+}
+```
+
+Both registration and authentication ceremonies happen in two steps. The client makes an initial request to the server with some identifier. The code below uses the username, but the user handle works as well. After prompting the user for a unique username, the server constructs a `PublicKeyCredentialCreationOptions` for registration, which contains the user handle, the domain of the key, and a challenge token.
+
+#### Registration endpoints
```

Comment:
This section seems like it could use a sentence as an intro. Or remove the #### header.

Reply:
Replaced with a bold block